zipkin-storage/zipkin2_cassandra/src/main/java/zipkin2/storage/cassandra/SelectSpanNames.java

```diff
@@ -0,0 +1,81 @@
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package zipkin2.storage.cassandra;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.Statement;
+import com.datastax.driver.core.querybuilder.QueryBuilder;
+import com.google.auto.value.AutoValue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+import zipkin2.Call;
+import zipkin2.storage.cassandra.internal.call.AccumulateAllResults;
+import zipkin2.storage.cassandra.internal.call.ResultSetFutureCall;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static zipkin2.storage.cassandra.Schema.TABLE_SERVICE_SPANS;
+
+@AutoValue
+abstract class SelectSpanNames extends ResultSetFutureCall {
+  static class Factory {
+    final Session session;
+    final PreparedStatement preparedStatement;
+    final AccumulateNamesAllResults accumulateNamesIntoSet = new AccumulateNamesAllResults();
+
+    Factory(Session session) {
+      this.session = session;
+      this.preparedStatement = session.prepare(
+        QueryBuilder.select("span").from(TABLE_SERVICE_SPANS)
+          .where(QueryBuilder.eq("service", QueryBuilder.bindMarker("service")))
+          .limit(QueryBuilder.bindMarker("limit_"))
+      );
+    }
+
+    Call<List<String>> create(String serviceName) {
+      if (serviceName == null || serviceName.isEmpty()) return Call.emptyList();
+      String service = checkNotNull(serviceName, "serviceName").toLowerCase();
+
+      return new AutoValue_SelectSpanNames(session, preparedStatement, service)
+        .flatMap(accumulateNamesIntoSet);
+    }
+  }
+
+  static class AccumulateNamesAllResults extends AccumulateAllResults<List<String>> {
+    @Override protected Supplier<List<String>> supplier() {
+      return ArrayList::new;// TODO: list might not be ok due to not distinct
+    }
+
+    @Override protected BiConsumer<Row, List<String>> accumulator() {
+      return (row, list) -> list.add(row.getString("span"));
+    }
+  }
+
+  abstract PreparedStatement preparedStatement();
+
+  abstract String service();
+
+  @Override public Statement newStatement() {
+    return preparedStatement().bind()
+      .setString("service", service())
+      .setInt("limit_", 1000); // no one is ever going to browse so many span names
+  }
+
+  @Override public SelectSpanNames clone() {
+    return new AutoValue_SelectSpanNames(session(), preparedStatement(), service());
```

Comment:
I will probably un-autovalue some or most of these as I need to override toString anyway as preparedStatement's toString is crap